mtp_api/apps/prison/forms.py

```diff
@@ -1,6 +1,82 @@
+import csv
+import decimal
+import io
+
 from django import forms
+from django.contrib.admin.widgets import AdminFileWidget
+from django.db import transaction
+from django.forms.models import ModelChoiceField
 from django.utils.translation import gettext_lazy as _
 
+from prison.models import Prison, PrisonerBalance, validate_prisoner_number
+
 
 class LoadOffendersForm(forms.Form):
     modified_only = forms.BooleanField(label=_('Only load updated offender locations'), required=False)
+
+
+class PrisonerBalanceUploadForm(forms.Form):
+    csv_file = forms.FileField(label=_('Prisoner balances file'), allow_empty_file=True, widget=AdminFileWidget)
+    prison = ModelChoiceField(queryset=Prison.objects.filter(use_nomis_for_balances=False))
+
+    def clean_csv_file(self):
+        csv_file = self.cleaned_data.get('csv_file')
+        if csv_file:
+            try:
+                return self.parse_balances(csv_file)
+            except Exception as e:
+                self.add_error('csv_file', f'Could not parse file: {e}')
+        return None
+
+    @classmethod
+    def parse_balances(cls, csv_file):
+        csv_contents = csv_file.read().decode()
+        csv_contents = io.StringIO(csv_contents)
+        csv_reader = csv.DictReader(csv_contents)
+        balances = []
+        for line in csv_reader:
+            prisoner_number = line['prisonernumber'].strip().upper()
+            # check that prisoner number follows expected pattern (else indicates malformed file)
+            validate_prisoner_number(prisoner_number)
+            amount = line['totalamount']
+            amount_pence = decimal.Decimal(amount.strip()) * 100
+            # check that amount_pence is not negative (assumed impossible)
+            # and has no decimal places (indicates malformed file)
+            numerator, denomiator = amount_pence.as_integer_ratio()
+            if numerator <= 0:
+                raise ValueError(f'Negative balance {amount}')
+            if denomiator != 1:
+                raise ValueError(f'Cannot turn {amount} into pence')
+            balances.append({
+                'prisoner_number': prisoner_number,
+                'amount': int(amount_pence),
+            })
+        return balances
+
+    @transaction.atomic
+    def save(self):
```

Comment:
It seems a bit strange to persist ORM objects on a form, maybe some kind of thin Serializer could be used?

Reply:
this is how the other file-uploading view does it. might as well use the same pattern, no? pretty sure that many forms in various apps actually do the db/api persistence inside them